[FLINK-30397][Connector/Pulsar] Remove Pulsar connector

MartijnVisser opened the pull request:
## What is the purpose of the change

The Pulsar connector code has been moved to https://github.com/apache/flink-connector-pulsar. 
* The `3.0` branch contains the 1.16 code, which has been officially released.
* The current `main` branch contains the code from this repo's `master` branch. 

I believe we can already remove the code from the repository, because the current code from `master` (or `main` if you look at the externalized connector repo) is a breaking change compared to the Flink 1.16 release and can only be released when Flink 1.17 has been released. Given that circumstances plus the instability of the Pulsar connector on the overall Flink CI stability would make it good to already remove the connector. 

## Brief change log

* Removed all source code, documentation and references
* Updated the Python code to refer to the externalized connector version

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

zentol commented on flink-docs/src/main/java/org/apache/flink/docs/util/ConfigurationOptionLocator.java:
```diff
@@ -73,15 +73,6 @@ public class ConfigurationOptionLocator {
                 new OptionsClassLocation("flink-clients", "org.apache.flink.client.cli"),
                 new OptionsClassLocation(
                         "flink-table/flink-sql-client", "org.apache.flink.table.client.config"),
-                new OptionsClassLocation(
-                        "flink-connectors/flink-connector-pulsar",
-                        "org.apache.flink.connector.pulsar.common.config"),
-                new OptionsClassLocation(
-                        "flink-connectors/flink-connector-pulsar",
-                        "org.apache.flink.connector.pulsar.source"),
-                new OptionsClassLocation(
-                        "flink-connectors/flink-connector-pulsar",
-                        "org.apache.flink.connector.pulsar.sink"),
```
The associated generated files should be removed. The pulsar docs will have to provide equivalent files _somehow_.
(medium-term we need to make the config docs generator available to connector repos)

MartijnVisser replied:
I would think that right now we should 'just' remove it here and create a follow-up ticket for a permanent solution, WDYT?

MartijnVisser replied:
> The pulsar docs will have to provide equivalent files somehow.

Can we tackle that as part of a follow-up ticket?

XComp replied:
Did we create the follow-up ticket?

MartijnVisser replied:
It is now in FLINK-30486

zentol commented on mvnw:
```diff
@@ -1,316 +0,0 @@
-#!/bin/sh
-# ----------------------------------------------------------------------------
-# Licensed to the Apache Software Foundation (ASF) under one
```
This shouldnt be deleted?

MartijnVisser replied:
Honestly don't know why that happened, reverted

zentol commented on tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/JarFileChecker.java:
```diff
@@ -210,11 +210,6 @@ private static int findNonBinaryFilesContainingText(
                     // dual-licensed under GPL 2 and EPL 2.0
                     // contained in sql-avro-confluent-registry
                     .filter(path -> !pathStartsWith(path, "/org/glassfish/jersey/internal"))
-                    // contained in sql-connector-pulsar
-                    .filter(
-                            path ->
-                                    !pathStartsWith(
-                                            path, "/org/apache/pulsar/shade/org/glassfish/jersey/"))
```
We need to keep this for now, as otherwise CI will fail in the pulsar repo (eventually).

zentol reviewed (changes requested):
you should update the production tests archunit store to remove pulsar violations.

MartijnVisser commented:
> you should update the production tests archunit store to remove pulsar violations.

Aren't the Pulsar violations already removed, since I've completely removed the Pulsar source code? I don't see any other violations in the `flink-architecture-tests-production` violations folder?

XComp commented on docs/setup_docs.sh:
```diff
@@ -42,6 +42,10 @@ function integrate_connector_docs {
   theme_dir="../themes/connectors"
   mkdir -p "${theme_dir}"
   rsync -a flink-connector-${connector}/docs/content* "${theme_dir}/"
+  DIR=flink-connector-${connector}/docs/layouts;
+  if [ -e $DIR ];
+    then rsync -a flink-connector-${connector}/docs/layouts* "${theme_dir}/"
+  fi
```
Why is this necessary now and wasn't necessary with the other externalized connectors?

MartijnVisser replied:
Because the Pulsar connector has generated shortcodes for the documentation, which no other externalized connector has had (so far). See https://github.com/apache/flink-connector-pulsar/tree/main/docs/layouts/shortcodes

XComp replied:
Shouldn't we make this then a parameter instead of hardcoding it in the function?
```
function integrate_connector_docs {
  local connector ref additional_folders
  connector=$1
  ref=$2
  additional_folders=( "${@[@]:3}" )

  git clone --single-branch --branch "${ref}" "https://github.com/apache/flink-connector-${connector}"
  theme_dir="../themes/connectors"
  mkdir -p "${theme_dir}"

  for rsync_folder_subpath in "docs/content*" "${additional_folders[@]}"; do
    local rsync_source_path="flink-connector-${connector}/${rsync_folder_subpath}"
    [ -e "${rsync_source_path}" ] || echo "Error: Invalid sub-path ${rsync_folder_subpath}! Folder does not exist." && exit 1

    rsync -a "flink-connector-${connector}/${rsync_folder_subpath}" "${theme_dir}/"
  done
}

# Integrate the connector documentation

rm -rf themes/connectors/*
rm -rf tmp
mkdir tmp
cd tmp

integrate_connector_docs elasticsearch v3.0.0
integrate_connector_docs aws v3.0.0
integrate_connector_docs cassandra v3.0.0
integrate_connector_docs pulsar main "docs/layouts*"
# [...]
```

MartijnVisser replied:
I have no strong opinion on this, so I'll take your suggestion

XComp reviewed (changes requested):
There's still the `JarChecker` that holds some pulsar reference. I guess, that's something we could get rid off as well? I assume that we want to keep the pyflink Pulsar-related source code.

MartijnVisser commented on docs/setup_docs.sh:
```diff
@@ -54,6 +58,7 @@ cd tmp
 integrate_connector_docs elasticsearch v3.0.0
 integrate_connector_docs aws v3.0.0
 integrate_connector_docs cassandra v3.0.0
+integrate_connector_docs pulsar main
```
No, because the externalized Pulsar connector in `main` is only compatible with Flink 1.17 (because of FLINK-28853) so it can only be updated when Flink 1.17 is released and there has been a new Pulsar release created (v4.0.0)

XComp replied:
I see, thanks for clarification.

MartijnVisser commented:
> There's still the `JarChecker` that holds some pulsar reference. I guess, that's something we could get rid off as well? I assume that we want to keep the pyflink Pulsar-related source code.

I've tried that, but PyFlink requires that actually.

XComp reviewed (commented):
I added a proposal to one of the thread. See [my comment above](https://github.com/apache/flink/pull/21501#discussion_r1054411775). About the `JarChecker`: Could we extend the comment to the JarChecker for the pulsar connector mentioning that it's required for flink-python?

MartijnVisser commented:
> I added a proposal to one of the thread. See [my comment above](https://github.com/apache/flink/pull/21501#discussion_r1054411775). 

Fixed

> About the `JarChecker`: Could we extend the comment to the JarChecker for the pulsar connector mentioning that it's required for flink-python?

Added :)

XComp commented on tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/JarFileChecker.java:
```diff
@@ -211,6 +211,7 @@ private static int findNonBinaryFilesContainingText(
                     // contained in sql-avro-confluent-registry
                     .filter(path -> !pathStartsWith(path, "/org/glassfish/jersey/internal"))
                     // contained in sql-connector-pulsar
+                    // while the Pulsar connector is externalized, this is still needed for PuFlink
```
```suggestion
                    // while the Pulsar connector is externalized, this is still needed for PyFlink
```

XComp commented on docs/setup_docs.sh:
```diff
@@ -36,14 +36,25 @@ echo "Created temporary file" $goModFileLocation/go.mod
   mkdir -p "${theme_dir}"
-  rsync -a flink-connector-${connector}/docs/content* "${theme_dir}/"
+
+  for rsync_folder_subpath in "docs/content*" "${additional_folders[@]}"; do
```
```suggestion
  for rsync_folder_subpath in "docs/content" "docs/content.zh" "${additional_folders[@]}"; do
```
sorry; my bad. The if statement doesn't allow the asterix. AFAIU, the asterix was just used initially to rsync both `docs/content` and `docs/content.zh`.

MartijnVisser replied:
This was also sloppy on my part, should have tried it locally first 😅

XComp commented on docs/setup_docs.sh:
```diff
@@ -54,6 +65,7 @@ cd tmp
 integrate_connector_docs elasticsearch v3.0.0
 integrate_connector_docs aws v3.0.0
 integrate_connector_docs cassandra v3.0.0
+integrate_connector_docs pulsar main "docs/layouts*"
```
```suggestion
integrate_connector_docs pulsar main "docs/layouts"
```
I know that just blindly copied the `layouts*`. So, it's actually my mistake. Sorry for that. But am I right in assuming that `docs/layouts*` was just copied from `docs/content*` without a reason? Or why do we need the asterix here? It appears that there is only a `docs/layout` subfolder in [flink-connector-pulsar:./docs/](https://github.com/apache/flink-connector-pulsar/tree/main/docs).

MartijnVisser replied:
> But am I right in assuming that docs/layouts* was just copied from docs/content* without a reason? 

Correct :) Fixed

XComp reviewed (commented):
I added some comments. PTAL

MartijnVisser commented:
@flinkbot run azure

XComp reviewed (approved):
LGTM now. Thanks for addressing my comments.
